### ISSUE TITLE
Add tape-aware fused/flash scaled-dot-product attention backward for training

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#889]
## Problem

AiDotNet training currently falls back to the materialized scaled-dot-product-attention path because the optimized float/FlashAttention branch is inference-only. Paper-scale transformer training therefore retains an attention-weight tensor for every block and can exceed a 16 GB runner even with serialized tests and streaming gradients.

This was reproduced while investigating `ooples/AiDotNet#1789` with Linux Docker constrained to 4 CPUs and 16 GB RAM. A paper-scale FP32 Bark training step failed in:

```text
System.OutOfMemoryException
  TensorBase<T>.GetFlattenedData()
  CpuEngine.ScaledDotProductAttentionBackward(...)
  BackwardFunctions<T>.ScaledDotProductAttentionBackward(...)
  GradientTape<T>.ComputeGradientsStreaming(...)
  NeuralNetworkBase<T>.TrainWithTapeStreaming(...)
  Bark<T>.Train(...)
```

A separate failure occurred during forward attention materialization through `Tensor<T>.Contiguous()`.

## Current limitation

`MultiHeadAttentionLayer<T>` can select the optimized float attention path only when it is not in training mode. Training calls the generic `ScaledDotProductAttention` implementation and saves/materializes `[batch, heads, sequence, sequence]` attention weights for backward.

For Bark defaults this repeats across the text encoder and codec LM transformer stacks. Complexity is expected to be quadratic in sequence length, but retaining/materializing the complete score/probability matrices across all blocks is avoidable.

## Requested implementation

Add a tape-aware fused or blockwise scaled-dot-product-attention operation with a matching backward implementation that:

- Supports FP32 CPU training first; preserve the generic fallback for unsupported types/configurations.
- Computes forward attention blockwise/online without materializing the full score matrix.
- Recomputes or stores only the minimal row statistics needed for backward.
- Produces gradients for query, key, and value compatible with `GradientTape<T>` and streaming-gradient callbacks.
- Supports causal and supplied masks, scaling, and the existing soft-cap behavior, or explicitly falls back when a feature is unsupported.
- Handles strided/permuted Q/K/V views without three unconditional full-tensor copies.
- Does not use the inference-only branch while a tape is recording unless its backward function is registered.

## Acceptance criteria

- Numerical forward and Q/K/V gradient parity against the existing implementation on small deterministic tensors, including finite-difference coverage.
- Tests for causal masking, external masking, scale, non-contiguous/permuted inputs, and supported soft-cap behavior.
- Peak-memory benchmark demonstrating that attention training memory no longer grows as a retained full `[B,H,S,S]` tensor per transformer block.
- End-to-end paper-scale transformer/Bark training step completes within a 16 GB Linux cgroup without an attention forward/backward OOM.
- Benchmark reports wall time and allocations versus the current CPU implementation for representative sequence lengths.

## Related performance observation

AiDotNet streaming training with exact gradient clipping currently performs two backward traversals. A memory-efficient attention backward will materially reduce both traversals, though eliminating the second traversal is a separate AiDotNet concern.
